soroban-sdk/src/contract_event.rs

```diff
@@ -24,43 +32,91 @@ use crate::{env::internal, Env, IntoVal, RawVal, Vec};
 /// # #[cfg(feature = "testutils")]
 /// # fn main() {
 /// #     let env = Env::default();
-/// #     let contract_id = FixedBinary::from_array(&env, [0; 32]);
+/// #     let contract_id = BytesN::from_array(&env, [0; 32]);
 /// #     env.register_contract(&contract_id, Contract);
 /// #     f::invoke(&env, &contract_id);
 /// # }
 /// # #[cfg(not(feature = "testutils"))]
 /// # fn main() { }
 /// ```
 #[derive(Clone)]
-pub struct ContractEvent(Env);
+pub struct Events(Env);
 
-impl Debug for ContractEvent {
+impl Debug for Events {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "ContractEvent")
+        write!(f, "Events")
+    }
+}
+
+pub trait IntoTopics {
+    fn into_topics(self, env: &Env) -> Object;
+}
+
+// 0 topics
+impl IntoTopics for () {
+    fn into_topics(self, env: &Env) -> Object {
+        Vec::<RawVal>::new(env).to_object()
     }
 }
 
-impl ContractEvent {
+macro_rules! impl_for_tuple {
+    ( $($typ:ident $idx:tt)* ) => {
+        impl<$($typ),*> IntoTopics for ($($typ,)*)
+        where
+            $($typ: IntoVal<Env, RawVal> + Clone),*
+        {
+            fn into_topics(self, env: &Env) -> Object {
+                $({
+                    let ev = self.$idx.clone().into_env_val(env);
+                    match TryInto::<EnvObj>::try_into(ev) {
+                        Ok(obj) => {
+                            if obj.as_object().is_obj_type(ScObjectType::Vec) {
+                                panic!("topic cannot be a vec")
+                            } else if obj.as_object().is_obj_type(ScObjectType::Map) {
+                                panic!("topic cannot be a map")
+                            } else if obj.as_object().is_obj_type(ScObjectType::Bytes) {
+                                let bytes = unsafe { Bytes::unchecked_new(obj.clone()) };
+                                if bytes.len() > TOPIC_BYTES_LENGTH_LIMIT {
+                                    panic!("topic exceeds bytes length limit")
+                                }
+                            }
+                        }
+                        Err(_) => (),
```

Comment:
The validations here should happen host side. There is no error handling on the WASM side so I don't think we gain much by doing the checks here.

Reply:
Yeah this makes sense, I've removed it.